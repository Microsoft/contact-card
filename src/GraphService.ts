import { IPersonaProfile, ResolveFunc, RejectFunc } from "./Types";
import { GraphServiceAuthenticator } from "./GraphServiceAuthenticator";

export module GraphService {
    const photosCache: { [userId: string]: Promise<string> } = {};
    const profilesCache: { [userId: string]: Promise<IPersonaProfile> } = {};
    const managersCache: { [userId: string]: Promise<IPersonaProfile> } = {};
    const directsCache: { [userId: string]: Promise<IPersonaProfile[]> } = {};

    const profileFields = "id,displayName,jobTitle,mail,department,officeLocation,city,businessPhones,imAddresses,companyName";
    const graphBaseUrl = "https://graph.microsoft.com";


    export async function resolveProfile(emailOrId: string): Promise<IPersonaProfile> {
        if (!profilesCache[emailOrId]) {
            const profilePromise = resolveProfileInternal(emailOrId);
            profilesCache[emailOrId] = profilePromise;
            cacheByIdOrEmail(profilePromise, profilePromise, profilesCache);
        }
        return profilesCache[emailOrId];
    }


    export async function getPhotoUrl(emailOrId: string): Promise<string> {
        if (!photosCache[emailOrId]) {
            photosCache[emailOrId] = getPhotoUrlInternal(emailOrId);
            cacheByIdOrEmail(profilesCache[emailOrId], photosCache[emailOrId], photosCache);
        }
        return photosCache[emailOrId];
    }


    export async function getManager(emailOrId: string): Promise<IPersonaProfile> {
        if (!managersCache[emailOrId]) {
            const mngrProfilePromise = getManagerInternal(emailOrId);
            managersCache[emailOrId] = mngrProfilePromise;

            cacheByIdOrEmail(profilesCache[emailOrId], mngrProfilePromise, managersCache);
            cacheByIdOrEmail(mngrProfilePromise, mngrProfilePromise, profilesCache);
        }
        return managersCache[emailOrId];
    }


    export async function getAllManagers(emailOrId: string): Promise<IPersonaProfile[]> {
        const res: IPersonaProfile[] = [];
        const maxDepth = 15;

        try {
            for (let i = 0; i < maxDepth; ++i) {
                const manager = await getManager(emailOrId);
                res.push(manager);
                emailOrId = manager.id;
            }
        } catch (e) {
            if (e.name !== "NotFound") {
                throw e;
            }
        }

        return res;
    }


    export async function getDirects(emailOrId: string): Promise<IPersonaProfile[]> {
        if (!directsCache[emailOrId]) {
            const directsPromise = getDirectsInternal(emailOrId);
            directsCache[emailOrId] = directsPromise;

            cacheByIdOrEmail(profilesCache[emailOrId], directsPromise, directsCache);
            directsPromise
                .then(directs => {
                    for (const direct of directs) {
                        cacheByIdOrEmail(Promise.resolve(direct), Promise.resolve(direct), profilesCache);
                    }
                })
                .catch(() => { /* nothing to cache */ });
        }
        return directsCache[emailOrId];
    }


    function cacheByIdOrEmail<T>(profilePromise: Promise<IPersonaProfile>, value: Promise<T>, cacheStorage: { [userId: string]: Promise<T> }) {
        if (!profilePromise) {
            return;
        }

        profilePromise
            .then(profile => {
                cacheStorage[profile.id] = value;
                if (profile.email) {
                    cacheStorage[profile.email] = value;
                }
            })
            .catch(() => { /* don't cache errors */ });
    }


    async function resolveProfileInternal(emailOrId: string): Promise<IPersonaProfile> {
        const response = await enqueueRequest(`/users/${emailOrId}?$select=${profileFields}`, "GET");
        if (response.status < 400) {
            return buildProfile(response.body);
        } else {
            throw Error((response.body && response.body.error && response.body.error.message) || response.status);
        }
    }

    // tslint:disable-next-line:no-any
    type BatchResponse = { id: string; status: number; headers: any; body: any };
    let queue: { [id: string]: { id: string; url: string; method: string; onResolve: ResolveFunc<BatchResponse>; onReject: RejectFunc } } = {};
    let lastRequestId = 0;
    let batchTimer: number | undefined;

    async function enqueueRequest(relativeUrl: string, method: string): Promise<BatchResponse> {
        ++lastRequestId;

        let onResolve: ResolveFunc<BatchResponse> = () => { /**/ };
        let onReject: RejectFunc = () => { /**/ };
        // tslint:disable-next-line:promise-must-complete -- will be resolved in a separate function
        const p = new Promise<BatchResponse>((resolve, reject) => {
            onResolve = resolve;
            onReject = reject;
        });

        queue[`${lastRequestId}`] = { id: `${lastRequestId}`, url: relativeUrl, method, onResolve, onReject };
        if (batchTimer) {
            clearTimeout(batchTimer);
        }
        batchTimer = window.setTimeout(async () => { await processBatch(); }, 0);

        return p;
    }


    async function processBatch() {
        let batchNumber = 0;
        const batchLimit = 20;
        const batches = {
            0: []
        };
        const processingQueue = queue;
        queue = {};

        // Create a dictionary mapping indices to arrays of 20 requests each
        for (const id of Object.keys(processingQueue)) {
            const req = processingQueue[id];

            if (batches[batchNumber].length >= batchLimit) {
                batchNumber += 1;
                batches[batchNumber] = [];
            }

            batches[batchNumber].push({
                id: req.id,
                method: req.method,
                url: req.url
            });
        }

        // Process each batch of 20 requests one at a time
        for (const current of Object.keys(batches)) {
            const requests = batches[current];
            const adToken = await GraphServiceAuthenticator.getAuthToken();
            const batchRequest = new Request(`${graphBaseUrl}/v1.0/$batch`, {
                method: "POST",
                headers: {
                    Authorization: `Bearer ${adToken}`,
                    "Content-Type": "application/json"
                },
                body: JSON.stringify({ requests })
            });

            try {
                const batchResponse = await fetch(batchRequest);
                if (batchResponse.ok) {
                    const data = await batchResponse.json();
                    for (const response of <BatchResponse[]>data.responses) {
                        processingQueue[response.id].onResolve(response);
                    }
                } else {
                    throw buildErrorFromResponse(batchResponse);
                }
            } catch (e) {
                // reject all
                for (const id of Object.keys(processingQueue)) {
                    processingQueue[id].onReject(e);
                }
            }
        }
    }


    async function getManagerInternal(emailOrId: string): Promise<IPersonaProfile> {
        const adToken = await GraphServiceAuthenticator.getAuthToken();
        const request = new Request(`${graphBaseUrl}/v1.0/users/${emailOrId}/manager?$select=${profileFields}`, {
            method: "GET",
            headers: { Authorization: `Bearer ${adToken}` }
        });

        const response = await fetch(request);
        if (response.ok) {
            const data = await response.json();
            return buildProfile(data);
        } else {
            throw buildErrorFromResponse(response);
        }
    }


    async function getPhotoUrlInternal(emailOrId: string): Promise<string> {
        const adToken = await GraphServiceAuthenticator.getAuthToken();
        const request = new Request(`${graphBaseUrl}/v1.0/users/${emailOrId}/photo/$value`, {
            method: "GET",
            headers: { Authorization: `Bearer ${adToken}` }
        });

        const response = await fetch(request, { cache: "force-cache" });
        if (response.ok) {
            const photoData = await response.blob();
            return URL.createObjectURL(photoData);
        } else {
            throw buildErrorFromResponse(response);
        }
    }


    async function getDirectsInternal(emailOrId: string): Promise<IPersonaProfile[]> {
        const adToken = await GraphServiceAuthenticator.getAuthToken();
        const request = new Request(`${graphBaseUrl}/v1.0/users/${emailOrId}/directReports?$select=${profileFields},accountEnabled`, {
            method: "GET",
            headers: { Authorization: `Bearer ${adToken}` }
        });

        const response = await fetch(request);
        if (response.ok) {
            const data = await response.json();
            return data.value
                // tslint:disable-next-line:no-any
                .filter((p: any) => p.accountEnabled !== false)
                .map(buildProfile);
        } else {
            throw buildErrorFromResponse(response);
        }
    }


    // tslint:disable-next-line:no-any
    function buildProfile(data: any): IPersonaProfile {
        return {
            id: data.id,
            displayName: data.displayName,
            jobTitle: data.jobTitle,
            department: data.department,
            email: data.mail,
            officeLocation: data.officeLocation,
            city: data.city,
            businessPhone: (data.businessPhones && data.businessPhones.length) ? data.businessPhones[0] : undefined,
            imAddress: (data.imAddresses && data.imAddresses.length) ? data.imAddresses[0] : undefined,
            companyName: data.companyName
        };
    }


    function buildErrorFromResponse(response: Response) {
        const ex = new Error(response.statusText);
        if (response.status === 404) {
            ex.name = "NotFound";
        }
        return ex;
    }
}
