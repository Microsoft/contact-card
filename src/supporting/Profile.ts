import { IPersonaProfile } from "../Types";


export function buildProfileResponse(id: number) {
    return {
        id: `userId${id}`,
        displayName: `Display Name ${id}`,
        jobTitle: `Job Title ${id}`,
        department: `Department ${id}`,
        mail: `user${id}@contoso.com`,
        officeLocation: `1234-${id}`,
        city: "Redmond",
        businessPhones: [`555-123-456${id}`],
        imAddresses: [`im-user${id}@contoso.com`],
        companyName: "Contoso"
    };
}


export function buildProfile(id: number): IPersonaProfile {
    return {
        id: `userId${id}`,
        displayName: `Display Name ${id}`,
        jobTitle: `Job Title ${id}`,
        department: `Department ${id}`,
        email: `user${id}@contoso.com`,
        officeLocation: `1234-${id}`,
        city: "Redmond",
        businessPhone: `555-123-456${id}`,
        imAddress: `im-user${id}@contoso.com`,
        companyName: "Contoso"
    };
}
