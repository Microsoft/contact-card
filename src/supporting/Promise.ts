export async function delay(ms: number) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}

export async function waitNextTick() {
    return delay(0);
}
