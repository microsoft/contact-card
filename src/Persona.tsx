import * as React from "react";
import { Persona as PersonaOF, PersonaSize, IPersonaProps as IPersonaPropsFluent } from "@fluentui/react";
import { IPersonaProfile, PersonaShowMode, ShowModeFlags } from "./Types";
import { GraphService } from "./GraphService";

export interface IPersonaProps {
    id?: string;
    email?: string;
    displayName?: string;
    showMode: PersonaShowMode;
    size?: PersonaSize;
    className?: string;
    internalProps?: IPersonaPropsFluent;
}


interface IPersonaState {
    profile: IPersonaProfile;
    photoUrl?: string;
}


export class Persona extends React.Component<IPersonaProps, IPersonaState> {
    private _isMounted: boolean = false;


    constructor(props: IPersonaProps) {
        super(props);
        this.state = this.buildStateFromProps(props);
    }


    public componentDidMount() {
        this._isMounted = true;
        this.resolveProfile();
    }


    public componentWillUnmount() {
        this._isMounted = false;
    }


    public componentDidUpdate(prevProps: IPersonaProps) {
        if (prevProps.id !== this.props.id || prevProps.email !== this.props.email) {
            this.setState(this.buildStateFromProps(this.props), () => this.resolveProfile());
        }
    }


    public render(): JSX.Element {
        return (
            <PersonaOF
                className={this.props.className}
                size={this.props.size || PersonaSize.size40}
                text={this.state.profile.displayName}
                // tslint:disable-next-line:no-bitwise
                secondaryText={this.props.showMode & ShowModeFlags.Title ? this.state.profile.jobTitle : undefined}
                // tslint:disable-next-line:no-bitwise
                tertiaryText={this.props.showMode & ShowModeFlags.Department ? this.state.profile.department : undefined}
                imageUrl={this.state.photoUrl}
                imageAlt={this.state.profile.displayName}
                {...this.props.internalProps}
            />
        );
    }


    private buildStateFromProps(props: IPersonaProps): IPersonaState {
        if (!props.id && !props.email) {
            throw new Error("Either id or email is required");
        }

        return {
            profile: {
                id: (props.id || props.email) as string,
                email: props.email,
                displayName: props.displayName || this.extractAlias(props.email)
            }
        };
    }


    private resolveProfile() {
        GraphService.resolveProfile(this.state.profile.id)
            .then(profile => {
                if (this._isMounted) {
                    this.setState({ profile, photoUrl: this.state.photoUrl });
                }
            })
            .catch(() => { /* swallow*/ });

        // tslint:disable-next-line:no-floating-promises -- because of the final `then`
        GraphService.getPhotoUrl(this.state.profile.id)
            .catch(() => undefined)
            .then((photoUrl: string | undefined) => {
                if (this._isMounted) {
                    this.setState({ profile: this.state.profile, photoUrl });
                }
            });
    }


    private extractAlias(email: string | undefined): string | undefined {
        if (!email) {
            return email;
        }

        const p = email.indexOf("@");
        return (p !== -1) ? email.substr(0, p) : email;
    }
}

