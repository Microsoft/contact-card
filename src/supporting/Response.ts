import { buildProfileResponse } from "./Profile";

export function buildResponse(userId: number, status: number) {
    return new Response(JSON.stringify({
        responses: [
            {
                id: userId,
                body: buildProfileResponse(userId),
                status: status
            }
        ]
    }));
}
