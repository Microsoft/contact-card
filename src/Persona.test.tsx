import * as React from "react";
import * as Enzyme from "enzyme";
import { Persona } from "./Persona";
import { buildProfile } from "./supporting/Profile";
import { PersonaShowMode } from "./Types";
import { GraphService } from "./GraphService";
import { waitNextTick } from "./supporting/Promise";
import { PersonaSize } from "office-ui-fabric-react";

jest.mock("./GraphService");


afterEach(() => {
    jest.resetAllMocks();
});


test("renders Persona by Id", async () => {
    GraphService.resolveProfile = () => Promise.resolve(buildProfile(1));
    GraphService.getPhotoUrl = () => Promise.resolve("blob://pic1");

    const wrap = Enzyme.shallow(<Persona id="userId1" showMode={PersonaShowMode.NameOnly} />);
    await waitNextTick();
    expect(wrap).toMatchSnapshot();
});


test("renders Persona by Email", async () => {
    const wrap = Enzyme.shallow(<Persona email="userId1@contoso.com" showMode={PersonaShowMode.NameOnly} />);
    expect(wrap).toMatchSnapshot();
});


test("renders Persona with Title", async () => {
    GraphService.resolveProfile = () => Promise.resolve(buildProfile(1));
    GraphService.getPhotoUrl = () => Promise.resolve("blob://pic1");

    const wrap = Enzyme.shallow(<Persona id="userId1" showMode={PersonaShowMode.NameTitle} />);
    await waitNextTick();
    expect(wrap).toMatchSnapshot();
});


test("renders Persona with Title and Department", async () => {
    GraphService.resolveProfile = () => Promise.resolve(buildProfile(1));
    GraphService.getPhotoUrl = () => Promise.resolve("blob://pic1");

    const wrap = Enzyme.shallow(<Persona id="userId1" showMode={PersonaShowMode.NameTitleDepartment} />);
    await waitNextTick();
    expect(wrap).toMatchSnapshot();
});


test("renders Persona with extra props", async () => {
    const wrap = Enzyme.shallow(<Persona id="userId1" displayName="Initial Display Name 1" size={PersonaSize.size100} className="class1" showMode={PersonaShowMode.NameOnly} />);
    expect(wrap).toMatchSnapshot();
});


test("renders Persona fails if no Id or EMail", async () => {
    expect(() => Enzyme.shallow(<Persona showMode={PersonaShowMode.NameOnly} />))
        .toThrowError(/required/);
});


test("renders Persona with failed profile", async () => {
    GraphService.resolveProfile = () => Promise.reject(new Error());
    GraphService.getPhotoUrl = () => Promise.reject(new Error());

    const wrap = Enzyme.shallow(<Persona id="userId1" showMode={PersonaShowMode.NameOnly} />);
    await waitNextTick();
    expect(wrap).toMatchSnapshot();
});


test("re-renders Persona by Email", async () => {
    const wrap = Enzyme.shallow(<Persona email="userId1@contoso.com" showMode={PersonaShowMode.NameOnly} />);
    const render1 = wrap.html();

    wrap.setProps({ email: "userId2@contoso.com", showMode: PersonaShowMode.NameOnly });
    const render2 = wrap.html();

    expect(render1).not.toBe(render2);
});
