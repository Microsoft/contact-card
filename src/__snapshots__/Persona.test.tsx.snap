// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Persona by Email 1`] = `
<StyledPersonaBase
  imageAlt="userId1"
  size={12}
  text="userId1"
/>
`;

exports[`renders Persona by Id 1`] = `
<StyledPersonaBase
  imageAlt="Display Name 1"
  imageUrl="blob://pic1"
  size={12}
  text="Display Name 1"
/>
`;

exports[`renders Persona with Title 1`] = `
<StyledPersonaBase
  imageAlt="Display Name 1"
  imageUrl="blob://pic1"
  secondaryText="Job Title 1"
  size={12}
  text="Display Name 1"
/>
`;

exports[`renders Persona with Title and Department 1`] = `
<StyledPersonaBase
  imageAlt="Display Name 1"
  imageUrl="blob://pic1"
  secondaryText="Job Title 1"
  size={12}
  tertiaryText="Department 1"
  text="Display Name 1"
/>
`;

exports[`renders Persona with extra props 1`] = `
<StyledPersonaBase
  className="class1"
  imageAlt="Initial Display Name 1"
  size={15}
  text="Initial Display Name 1"
/>
`;

exports[`renders Persona with failed profile 1`] = `
<StyledPersonaBase
  size={12}
/>
`;
